Allow iam_role to work with friendly policy names

willthames opened the pull request:
##### SUMMARY

Basically obtain all the policies and create a mapping
from friendly name to ARN

`ParamValidationError` doesn't contain a response key
so needs different exception handling to `ClientError`

Fix problem with methods that call `module.fail_json`
but don't take a module argument

Supersedes #22618

##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
iam_role

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel ce56da69b2) last updated 2017/03/15 09:51:56 (GMT +1000)
  config file = /home/will/src/xvt-dpe-cloud-infrastructure-stream2/ansible.cfg
  configured module search path = [u'./library']
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION

wimnat reviewed (commented):
Nice!

I'll wait for this to get merged then I have a fix for module always reporting changed

wimnat commented on lib/ansible/modules/cloud/amazon/iam_role.py:
```diff
@@ -43,9 +43,10 @@
     description:
-      - A list of managed policy ARNs (can't use friendly names due to AWS API limitation) to attach to the role. To embed an inline policy,
-        use M(iam_policy). To remove existing policies, use an empty list item.
+      - A list of managed policy ARNs or friendly names to attach to the role. To embed an inline policy, use M(iam_policy).
```
I would make a note that < 2.4, only ARNs work

wimnat commented on lib/ansible/modules/cloud/amazon/iam_role.py:
```diff
@@ -180,13 +181,29 @@ def compare_attached_role_policies(current_attached_policies, new_attached_polic
+    allpolicies = {}
+    paginator = connection.get_paginator('list_policies')
+
+    for response in paginator.paginate(PaginationConfig=dict(MaxItems=1000)):
```
I would use `build_full_result()` rather than arbitrary MaxItems.

E.g.
`paginator.paginate().build_full_result()`

wimnat commented on lib/ansible/modules/cloud/amazon/iam_role.py:
```diff
@@ -336,16 +352,13 @@ def main():
             path=dict(default="/", required=False, type='str'),
             assume_role_policy_document=dict(required=False, type='json'),
-            managed_policy=dict(default=None, required=False, type='list'),
+            managed_policy=dict(default=[], required=False, type='list', aliases=['managed_policies']),
```
Default needs to remain None

wimnat commented on lib/ansible/modules/cloud/amazon/iam_role.py:
```diff
@@ -180,13 +181,29 @@ def compare_attached_role_policies(current_attached_policies, new_attached_polic
     params['RoleName'] = module.params.get('name')
     params['AssumeRolePolicyDocument'] = module.params.get('assume_role_policy_document')
-    managed_policies = module.params.get('managed_policy')
+    managed_policies = convert_friendly_names_to_arns(connection, module, module.params.get('managed_policy'))
```
Only do this is module.params.get('managed_policy') is not None

Commit:
Allow iam_role to work with friendly policy names

Basically obtain all the policies and create a mapping
from friendly name to ARN

Tidy up module argument specification to remove unneeded
settings.

willthames commented:
@wimnat I believe I've addressed all of your feedback now, please rereview.

wimnat commented:
shipit

willthames commented:
!needs_revision

willthames commented:
@wimnat can you approve this PR too to remove the needs_revision tag

willthames commented:
bot_status

ryansb commented on lib/ansible/modules/cloud/amazon/iam_role.py:
```diff
@@ -180,13 +181,31 @@ def compare_attached_role_policies(current_attached_policies, new_attached_polic
+        return policy_names
+    allpolicies = {}
+    paginator = connection.get_paginator('list_policies')
+    policies = paginator.paginate().build_full_result()['Policies']
```
This API does have a rate limit, so it'd be good to add a backoff here. I'm not sure if boto3 paginators handle that automatically.
